### DISCUSSION TITLE
extractAttitude question

### DISCUSSION THREAD

**Discussion by blueRaining** [heremaps/harp.gl#2253] [Q&A]
this is MapViewUtils.extractAttitude 

if (d < 1.0 - Number.EPSILON) {
            if (d > -1.0 + Number.EPSILON) {
                yaw = Math.atan2(space.z.x, -space.z.y);
                pitch = Math.acos(space.z.z);
              //**************************HERE**************************************************
                roll = Math.atan2(space.x.z, space.y.z);
            //****************************************************************************
            } else {
                // Looking bottom-up with space.z.z == -1.0
                yaw = -Math.atan2(-space.y.x, space.x.x);
                pitch = 180;
                roll = 0;
            }
        }
this is MapView.extractAttitude 

if (d < 1.0 - epsilon) {
            if (d > -1.0 + epsilon) {
                yaw = Math.atan2(transform.zAxis.x, -transform.zAxis.y);

                pitch = Math.acos(transform.zAxis.z);
           //****************************HERE********************************************
                roll = Math.atan2(transform.xAxis.x, transform.yAxis.z);
       //************************************************************************
            } else {
                // Looking bottom-up with space.z.z == -1.0
                yaw = -Math.atan2(-transform.yAxis.x, transform.xAxis.x);
                pitch = 180;
                roll = 0;
            }
        }

roll is different
can we use camera euler to replace it?

**Comment by nzjony** [comment]
@blueRaining , it seems that one of them is wrong. The roll value in the function in the MapView is just used for the `MapViewEventNames.CameraPositionChanged` event, and not in any code, so my guess is that this code is incorrect.

What do you mean by `can we use camera euler to replace it?` ?

**Reply by blueRaining** [reply]
thank you for your replay,can we use camera.getWorldQuaternion and use Euler.setFromQuaternion to get camera world rotation replace compute by extractAttitude

**Reply by blueRaining** [reply]
i like harpgl very much,and i have read the source code a longtime  ,it's my pleasure to get more detail about harpgl

**Reply by nzjony** [reply]
Should work, as long as you transform the camera position to be relative to some position, i.e. for example, the extractAttitude function in MapView computes the localTangentSpace of the `m_targetGeoPos`, i.e. the target of the camera.
